src/Microsoft.AspNetCore.Routing/Matching/Ascii.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.AspNetCore.Routing.Matching
+{
+    internal static class Ascii
+    {
+        // case-sensitive equality comparison when we KNOW that 'a' is in the ASCII range
+        // and we know that the spans are the same length.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool UnsafeAsciiIgnoreCaseEqual(ReadOnlySpan<char> a, ReadOnlySpan<char> b, int length)
+        {
+            // We're doing unsafe operations here, caller is expected to have already checked
```

Comment:
I always worry when I see the caller is "expected to have already checked", are we expecting anything outside of code you write in routing to end up calling this? And is the expectation correct everywhere right now?

Reply:
What happens if `ref Unsafe.Add(ref charA, 1);` reads beyond the end of charA? An exception? Reading random memory?

Reply:
@blowdart - this is new code that's called in two specific places in routing. 

https://github.com/aspnet/Routing/pull/729/files#diff-fd3f742fd5ae89148383cc0f192ce283R43 - checked at lines: 38 & 42

https://github.com/aspnet/Routing/pull/729/files#diff-c9b9e0a59a91deab3cd4517572d902eeR51 - checked at lines: 41 & 48

We may end up using these more in routing, but always within the same family of internal string processing classes.

Would you feel better if I just inlined this code so it doesn't leak? 

--- 

@JamesNK - **UNDEFINED BEHAVIOUR** 😆

Reply:
@blowdart - to close the loop on this I've removed this comment as it's no longer true. We check the length here now.